Document Instructions and various network integrations

kayabaNerve opened the pull request:
Closes https://github.com/serai-dex/serai/issues/57.

Does use terminology from #53.

kayabaNerve committed:
Document Instructions and various network's integrations

Tracking issue: https://github.com/serai-dex/serai/issues/57

kayabaNerve committed:
Include origin as an Option in Shorthand

Converts (Network, Address) to Enum { Native(Address), Serai(Address) } 
as it's not valid to send Bitcoin to Ethereum.

Corrects a legacy comment regarding serialization.

kayabaNerve commented on docs/integrations/Bitcoin.md:
```diff
@@ -0,0 +1,23 @@
+  - `p2sh`:   20-byte hash.
+  - `p2wpkh`: 20-byte hash.
+  - `p2wsh`:  32-byte hash.
+  - `p2tr`:   32-byte key.
```
Bitcoin does not validate P2TR keys, so this is fine.

kayabaNerve commented on docs/integrations/Bitcoin.md:
```diff
@@ -0,0 +1,23 @@
+
+Bitcoin In Instructions are present via the transaction's last output in the
+form of `OP_RETURN`, and accordingly limited to 80 bytes. `origin` is
+automatically set to the transaction's first input's address, if recognized.
```
I also debated the change output. There's guaranteed to be a input though, and there's not guaranteed to be a change output. It will be more annoying to recover the input's address though.

TheArchitect108 replied:
Less = more in this case

kayabaNerve commented on docs/integrations/Ethereum.md:
```diff
@@ -0,0 +1,16 @@
+Ethereum In Instructions are present via being appended to the calldata
+transferring funds to Serai. `origin` is automatically set to the party from
+which funds are being transferred. For an ERC20, this is `from`. For ETH, this
+is the caller. `data` is limited to 255 bytes.
```
Matches Monero, fits in a single byte. A Curve trade takes ~120 bytes. Uv3 are massive at ~580 bytes, and likely best with an interim providing that expansion. No further considerations were made.

kayabaNerve commented on docs/integrations/Instructions.md:
```diff
@@ -0,0 +1,160 @@
+actual values.
+  - Vectors are prefixed by their length. If omitted, `vec![]`.
+  - Instruction fields are numbered and sequentially encoded, each prefixed by
+an ordinal byte. All other fields are sequentially encoded with no markers.
```
Instruction fields are more frequently omitted, hence why we solely mention the ones included.

Shorthand fields are more frequently included, hence why we don't bother numbering them.

kayabaNerve commented on docs/integrations/Instructions.md:
```diff
@@ -0,0 +1,160 @@
+### Out Instructions
+
+  - `destination` (Enum { Native(Address), Serai(Address) }): Address to receive
+funds to.
```
We *can* convert BTC origin addresses to ETH addresses, and can convert used ETH EOA accounts to BTC addresses. I don't believe any wallet would dare touch such functionality, and while it's interesting, and would get memos to 10 bytes, is best not implemented (not to mention the scope).

kayabaNerve replied:
This was originally (Network, Address). Sending XMR to an ETH address isn't valid though. The authenticated calling API could be used to build a wrapped token system, and is the intended path for any potential bridge. It'd be appropriately sandboxed and doesn't raise the question of XMR -> BTC (SLP tokens? I guess?), which would be a possible combination for simply (Network, Address).

kayabaNerve replied:
To elaborate, making any bridge a first class citizen doesn't really save on complexity, especially due to the scope creep. It would decrease gas costs on ETH, yet with increased first-party complexity when those are already a large burden, and this keeps thing modular. There is another discussion a bridge still needs to be a first class citizen to enable redemptions (/ we need a new API exposed for that), yet we're not tackling any bridge discussions yet (not even as a tracking issue, and I'd prefer to keep it that way). We're solely acknowledging API flexibility.

kayabaNerve committed:
Add authenticated calls to Ethereum

Also uses numbered lists for function descriptions.

kayabaNerve commented on docs/integrations/Instructions.md:
```diff
@@ -0,0 +1,160 @@
+  3) Checks the amount of `coin` received is greater than `minimum`.
+  4) Executes `out` with the amount of `coin` received.
+
+For a Bitcoin to Monero swap, Swap Shorthand is expected to generally take:
```
This may be better left to the PR comments, yet this does have to be less than 80 bytes. This math shows it is.

kayabaNerve commented on docs/integrations/Instructions.md:
````diff
@@ -0,0 +1,160 @@
+}
+```
+
+would cause the created seraiBTC to be transferred to the Bitcoin Asset Contract
````
sBTC is synth BTC, which is not what we have. sriBTC isn't optimal. I do prefer the long form seraiBTC, with a token symbol of simply BTC on Serai.

kayabaNerve commented on docs/integrations/Monero.md:
```diff
@@ -0,0 +1,24 @@
+### In Instructions
+
+Monero In Instructions are present via `tx.extra`, specifically via inclusion
+in a `TX_EXTRA_TAG_PADDING` tag, and accordingly limited to 255 bytes.
```
Monero wallets just error when scanning anything larger. We *could* ignore this without issue, yet we have no incentive to at this time, and it may introduce incompatibilities with wallets sending to us.

kayabaNerve commented on docs/integrations/Instructions.md:
```diff
@@ -0,0 +1,160 @@
+
+where `swap_and_add_liquidity` is a function which:
+
+  1) Swaps half of the incoming funds for SRI.
```
This actually leaves a potentially notable amount of potential liquidity on the table, as it doesn't consider slippage at all. It's just also trivial to calculate, and the amounts aren't notable at all when slippage isn't notable due to the size of the liquidity pool.

kayabaNerve commented on docs/integrations/Instructions.md:
```diff
@@ -0,0 +1,160 @@
+
+Or 74 bytes.
+
+##### Add Liquidity
```
Ideally, Add Liq would solely be from Serai. The issue is the Monero UX.

Monero can:

- Swap and add (1 XMR TX, 108 bytes)
- Swap for Serai + gas, wrap, add liquidity (1 XMR TX, 107 bytes, wait 20 minutes, 1 XMR TX, ~98 bytes, 1 Serai TX)
- Swap for Serai while also wrapping (1 XMR TX, 111 bytes, 1 Serai TX)

So I believe we have to accept this exposure and 108 bytes on Monero, not including the 1 byte label included on Monero (and tx.extra VarInt length size increase, +1 byte there too).

TheArchitect108 commented on docs/integrations/Instructions.md:
```diff
@@ -0,0 +1,160 @@
+Instructions are interpreted according to their non-Serai network. Addresses
+have no validation performed, beyond being a valid enum entry (when applicable)
+of the correct length, unless otherwise noted. If the processor is instructed to
+act on invalid data, or send to itself, it will drop the entire instruction.
```
We should document refunds.

kayabaNerve replied:
Also fee logic, gas limits, how we charge fees for USDC transfers out... yet out of scope for this immediate.

TheArchitect108 commented on docs/integrations/Instructions.md:
```diff
@@ -0,0 +1,160 @@
+appropriate amount of coins, and transferring them to the specified target with
+the attached data.
+
+If the transaction fails, funds are scheduled to be returned to `origin`.
```
transaction OR instruction?

kayabaNerve replied:
It's an equivalence here. Keeping things in terms of "instruction" is likely optimal though :)

kayabaNerve committed:
Document Serai's Application Calls and update Instructions accordingly

kayabaNerve committed:
Correct link in Instructions

kayabaNerve committed:
Update according to comment

kayabaNerve committed:
Document various Scenarios

- Pong
- Wrap
- SRI -> BTC
- BTC -> Monero
- Add Liquidity (fresh)
- Add Liquidity (SRI holder)
